jetty-core/jetty-compression/jetty-compression-server/src/main/java/org/eclipse/jetty/compression/server/CompressionHandler.java

```diff
@@ -0,0 +1,280 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.compression.server;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+import org.eclipse.jetty.compression.Compression;
+import org.eclipse.jetty.http.EtagUtils;
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.MimeTypes;
+import org.eclipse.jetty.http.pathmap.MatchedResource;
+import org.eclipse.jetty.http.pathmap.PathMappings;
+import org.eclipse.jetty.http.pathmap.PathSpec;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CompressionHandler extends Handler.Wrapper
+{
+    public static final String HANDLER_ETAGS = CompressionHandler.class.getPackageName() + ".ETag";
+
+    private static final Logger LOG = LoggerFactory.getLogger(CompressionHandler.class);
+    // TODO: make into a case-insensitive map
+    private final Map<String, Compression> supportedEncodings = new HashMap<>();
+    private final PathMappings<CompressionConfig> pathConfigs = new PathMappings<CompressionConfig>();
+
+    public void addCompression(Compression compression)
+    {
+        if (isRunning())
+            throw new IllegalStateException("Unable to add Compression on running DynamicCompressionHandler");
+
+        supportedEncodings.put(compression.getEncodingName(), compression);
+        compression.setContainer(this);
+        addBean(compression);
+    }
+
+    public CompressionConfig getConfiguration(PathSpec pathSpec)
+    {
+        // TODO: should we return null here?
+        //   It would make using this for additive changes to an existing configuration tricky (from an XML)
+        return pathConfigs.computeIfAbsent(pathSpec, (spec) ->
+        {
+            return CompressionConfig.builder().build();
+        });
+    }
+
+    public CompressionConfig getConfiguration(String pathSpecString)
+    {
+        PathSpec pathSpec = PathSpec.from(pathSpecString);
+        return getConfiguration(pathSpec);
+    }
+
+    public void putConfiguration(PathSpec pathSpec, CompressionConfig config)
```

Comment:
You may as well return the result of the put, so old values can be seen/checked.

Reply:
Also, whilst the impl uses a "put" I think this is probably better named as an "add".  

Also, it is a bit strange to have the method use "Configuration" when the interface is "CompressionConfig".  Try to be consistent.   Maybe have "CompressionHandler.Configuration" ?